### PR TITLE
Switch to PEP 517 packaging using hatchling

### PR DESCRIPTION
**Pull request by akx**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [ ] Does `$ tox` pass with this change (including linting)?
- [ ] Do the CI tests pass with this change (enable it first in your forked repo and wait for the github action build to finish)?
- [x] Is the new or changed code fully tested?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Is there an example added to the examples folder (if applicable)?
  - not applicable 
- [ ] Was the change added to CHANGES file?
  - not yet
 
### Description of change

This PR shifts redis-py from legacy `setup.py` packaging to [PEP 517](https://peps.python.org/pep-0517/) compliant packaging [using `hatchling` from PyPA](https://hatch.pypa.io/latest/).

Essentially, this does what was discussed in #1658 but with the increasingly standard `hatchling` tool instead of poetry's idiosyncracies.

Refs #1316 (which initially did this, with tools from that age)
Refs #1649 (which reverted that for no good reason)

I wasn't able to test all of these changes yet, but `python -m build` does create a working wheel with all modules included.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+keywords = [
+    "Redis",
+    "key-value store",
+    "database",
+]
+license = "MIT"
+version = "4.4.0rc1"
+url = "https://github.com/redis/redis-py"
+author = "Redis Inc."
+author_email = "oss@redis.com"
```
Comment (ofek) [inline pyproject.toml:13]:
```suggestion
authors = [
    { name = "Redis Inc.", email = "oss@redis.com" },
]
```

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+requires-python = ">=3.7"
+dependencies = [
+    "deprecated>=1.2.3",
+    "packaging>=20.4",
+    "importlib-metadata >= 1.0; python_version < \"3.8\"",
+    "typing-extensions; python_version<\"3.8\"",
```
Comment (ofek) [inline pyproject.toml:19]:
```suggestion
    "importlib-metadata >= 1.0; python_version < '3.8'",
    "typing-extensions; python_version<'3.8'",
```

**Comment by chayim** [comment]
With no clear winner in the packaging space (yet) - and the ongoing flux, I don't see redis-py moving from setup.py currently. As this changes, it too should change. But hatch, is yet another attempt at something, and we'll watch it as it gains steam.

**Comment by ofek** [comment]
> I don't see redis-py moving from setup.py currently.

Hey! Do you have some heuristic in mind, so the next person doesn't waste their time?

> the ongoing flux

I don't see much of that. The `pyproject.toml` way is for sure the standard, is [documented](https://packaging.python.org/en/latest/tutorials/packaging-projects/#creating-pyproject-toml) as such, and the execution of `setup.py` files is [deprecated](https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html).

**Comment by ischaojie** [comment]
sadly to see this :(, `pyproject.toml` is already the de facto standard.

**Pull request closed without merging**